core/lib/contract_verifier/src/compilers/vyper.rs

```diff
@@ -0,0 +1,115 @@
+use std::{collections::HashMap, mem, path::PathBuf, process::Stdio};
+
+use anyhow::Context;
+use tokio::io::AsyncWriteExt;
+use zksync_queued_job_processor::async_trait;
+use zksync_types::contract_verification_api::{
+    CompilationArtifacts, SourceCodeData, VerificationIncomingRequest,
+};
+
+use super::{parse_standard_json_output, process_contract_name, Settings, Source, StandardJson};
+use crate::{error::ContractVerifierError, resolver::Compiler};
+
+#[derive(Debug)]
+pub(crate) struct VyperInput {
+    pub contract_name: String,
+    pub file_name: String,
+    pub sources: HashMap<String, String>,
+    pub optimizer_mode: Option<String>,
+}
+
+impl VyperInput {
+    pub fn new(req: VerificationIncomingRequest) -> Result<Self, ContractVerifierError> {
+        let (file_name, contract_name) = process_contract_name(&req.contract_name, "vy");
+
+        let sources = match req.source_code_data {
+            SourceCodeData::VyperMultiFile(s) => s,
+            other => unreachable!("unexpected `SourceCodeData` variant: {other:?}"),
+        };
+        Ok(Self {
+            contract_name,
+            file_name,
+            sources,
+            optimizer_mode: if req.optimization_used {
+                // FIXME: values have differing semantics compared to other compilers; is this OK?
```

Comment:
Subj. It probably would make sense to make the optimizer mode dependent on the specified compilers on the type level, or at least check correspondence in the API.

Reply:
Probably; but the same as with above issue -- it works today, so, it's probably not that big of an issue.